contracts/FlowContractAudits.cdc

```diff
@@ -0,0 +1,215 @@
+pub contract FlowContractAudits {
+    
+    // Event that is emitted when a new Auditor resource is created
+    pub event AuditorCreated()
+
+    // Event that is emitted when a new contract audit voucher is created
+    pub event VoucherCreated(address: Address?, recurrent: Bool, expiryBlockHeight: UInt64?, codeHash: String)
+
+    // Event that is emitted when a contract audit voucher is used
+    pub event VoucherUsed(address: Address, key: String, recurrent: Bool, expiryBlockHeight: UInt64?)
+
+    // Event that is emitted when a contract audit voucher is removed
+    pub event VoucherRemoved(key: String, recurrent: Bool, expiryBlockHeight: UInt64?)
+
+    // Dictionary of all vouchers
+    access(contract) var vouchers: {String: AuditVoucher}
+
+    // The storage path for the admin resource
+    pub let AdminStoragePath: StoragePath
+
+    // The storage Path for auditors' AuditorProxy
+    pub let AuditorProxyStoragePath: StoragePath
+
+    // The public path for auditors' AuditorProxy capability
+    pub let AuditorProxyPublicPath: PublicPath
+
+    // Single audit voucher that is used for contract deployment
+    pub struct AuditVoucher {
+        
+        // Address of the account the voucher is intended for
+        // If nil, the contract can be deployed to any account
+        pub let address: Address?
+
+        // If false, the voucher will be removed after first use
+        pub let recurrent: Bool
+
+        // If non-nil, the voucher won't be valid after the expiry block height
+        pub let expiryBlockHeight: UInt64?
+        
+        // Hash of contract code
+        pub let codeHash: String
+        
+        init(address: Address?, recurrent: Bool, expiryBlockHeight: UInt64?, codeHash: String) {
+            self.address = address
+            self.recurrent = recurrent
+            self.expiryBlockHeight = expiryBlockHeight
+            self.codeHash = codeHash
+        }
+    }
+
+    // Returns all current vouchers
+    pub fun getAllVouchers(): {String: AuditVoucher} {
```

Comment:
Not sure what the use case of this method is? Is there a need for anyone to access the list of all vouchers?

It might be better to have a method where you can ask if there is a voucher for specific code and address.

On second thought... everything on chain is public, so I'm not sure if my concerns are valid. The only other concern is that if `self.vouchers` grows large this method will start failing...

Reply:
There are multiple mechanisms in place to make sure the dictionary on this contract is not only increasing:
* Non-recurrent vouchers are removed on use
* Vouchers optionally expire based on block height
* Admin can call `cleanupExpiredVouchers` to remove all the vouchers that are currently expired and haven't been used. 
There is no high priority use case with the ability to fetch all vouchers (I have a sample script that does this), but I was thinking there is no way to fetch them all and check/debug for something in the future, hence this function. 
I agree with your feedback that if the "current" voucher list grows (like 10k? maybe we define a lot of recurrent templates) there might be performance issues with fetching them all. This argument holds for any data a contract might return. 
For this case with all the above, I prefer to keep the function but let me know if I'm missing something.